platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java

```diff
@@ -22,51 +22,75 @@
 
 import static org.gradle.internal.execution.history.changes.ExecutionStateChanges.nonIncremental;
 
-public class ResolveChangesStep<C extends CachingContext, R extends Result> implements Step<C, R> {
+public class ResolveChangesStep<C extends ValidationFinishedContext, R extends Result> implements Step<C, R> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResolveChangesStep.class);
     private static final ImmutableList<String> NO_HISTORY = ImmutableList.of("No history is available.");
     private static final ImmutableList<String> UNTRACKED = ImmutableList.of("Change tracking is disabled.");
     private static final ImmutableList<String> VALIDATION_FAILED = ImmutableList.of("Incremental execution has been disabled to ensure correctness. Please consult deprecation warnings for more details.");
 
     private final ExecutionStateChangeDetector changeDetector;
+    private final BooleanSupplier emitBuildCacheDebugLogging;
 
     private final Step<? super IncrementalChangesContext, R> delegate;
 
     public ResolveChangesStep(
         ExecutionStateChangeDetector changeDetector,
+        BooleanSupplier emitBuildCacheDebugLogging,
         Step<? super IncrementalChangesContext, R> delegate
     ) {
         this.changeDetector = changeDetector;
+        this.emitBuildCacheDebugLogging = emitBuildCacheDebugLogging;
         this.delegate = delegate;
     }
 
     @Override
     public R execute(UnitOfWork work, C context) {
         IncrementalChangesContext delegateContext = context.getBeforeExecutionState()
             .map(beforeExecution -> resolveExecutionStateChanges(work, context, beforeExecution))
-            .map(changes -> new IncrementalChangesContext(context, changes.getChangeDescriptions(), changes))
+            .map(changes -> {
+                HashCode cacheKey = context.getPreviousExecutionState()
+                    .filter(__ -> changes.getChangeDescriptions().isEmpty())
+                    .map(PreviousExecutionState::getCacheKey)
+                    .orElseGet(() -> calculateCacheKey(changes));
+                return new IncrementalChangesContext(context, changes.getChangeDescriptions(), changes, cacheKey);
+            })
             .orElseGet(() -> {
                 ImmutableList<String> rebuildReason = context.getNonIncrementalReason()
                     .map(ImmutableList::of)
                     .orElse(UNTRACKED);
-                return new IncrementalChangesContext(context, rebuildReason, null);
+                return new IncrementalChangesContext(context, rebuildReason, null, null);
             });
 
         return delegate.execute(work, delegateContext);
     }
 
+    private HashCode calculateCacheKey(ExecutionStateChanges changes) {
+        Logger logger = emitBuildCacheDebugLogging.getAsBoolean()
+            ? LOGGER
+            : NOPLogger.NOP_LOGGER;
+        return new DefaultCachingStateFactory(logger).calculateCacheKey(changes.getBeforeExecutionState());
+    }
```

Comment:
:x: This should only receive the `BeforeExecutionState`.

Reply:
Done.